### DISCUSSION TITLE
Same algorithm (less than 10 LOC) in Rust is >10 times faster. How come?

### DISCUSSION THREAD

**Discussion by ghost** [modular/modular#1790] [General]
mojo:

```mojo
fn main() raises:
    with open("nouns.txt", "r") as f:
        let lines = f.read()
        var words = lines.split("\n")
        for word in range(len(words)):
            print_no_newline(words[word], "")
```
```bash
$ mojo build main.🔥
$ time ./main
...
real	0.845
mem	9KB
```

rust:

```rust
use std::io::Read;

fn main() {
  let mut file = std::fs::File::open("nouns.txt").unwrap();
  let mut words = String::new();
  let _ = file.read_to_string(&mut words);
  let words = words.split('\n');
  for word in words {
    print!("{word} ");
  }
}
```
```bash
$ cargo build --release
$ time ./target/release/main
...
real	0.072
mem	5KB
```

nouns.txt: 
[nouns.txt](https://github.com/modularml/mojo/files/14356659/nouns.txt)

arch linux
latest stable mojo (0.7.0 (af002202))
latest stable rust (1.76.0 (07dca489a 2024-02-04))
kitty terminal emulator

**Comment by ghost** [comment]
but when i print one word on a line the times are 0.820 and 0.211 respectively

**Comment by Sharktheone** [comment]
From my testing it looks like print calls in mojo aren't quite on the level rust's are. When I change the for loop to append to a DynamicVector it completes in 0.029 seconds. Probably Modular needs to improve the print functions

**Comment by MoSafi2** [comment]
Print calls and string operations in Mojo are still quite unoptimized, also the split line operation in Mojo as I remember from my experimentation was quite expensive. 
You can try the following

```python
fn main() raises:
    with open("nouns.txt", "r") as f:
        let lines = f.read_bytes() #Faster than f.read(), returns Tensor[Dtype.int8]
```
and write a custom function that loops over input looking for line seperator.
You can find here experiments with newline finding function using iteration or SIMD vectorization. 
https://github.com/MoSafi2/MojoFastTrim/blob/b5f60ccea2c66fa42c1b1e7b5ed7f267afbe720e/MojoFastTrim/helpers.mojo#L19
